ui/src/extensions/mod.rs

```diff
@@ -0,0 +1,74 @@
+use extensions::*;
+use libloading::Library;
+use std::{collections::HashMap, ffi::OsStr, io, rc::Rc};
+
+struct ExtensionRegistrar {
+    extensions: HashMap<String, ExtensionProxy>,
+    lib: Rc<Library>,
+}
+
+impl ExtensionRegistrar {
+    fn new(lib: Rc<Library>) -> ExtensionRegistrar {
+        ExtensionRegistrar {
+            lib,
+            extensions: HashMap::default(),
+        }
+    }
+}
+
+impl extensions::ExtensionRegistrar for ExtensionRegistrar {
+    fn register(&mut self, name: &str, extension: Box<dyn Extension>) {
+        let proxy = ExtensionProxy {
+            extension,
+            _lib: Rc::clone(&self.lib),
+        };
+        self.extensions.insert(name.to_string(), proxy);
+    }
+}
+
+#[derive(Default)]
+pub struct AvailableExtensions {
+    pub extensions: HashMap<String, ExtensionProxy>,
+    pub libraries: Vec<Rc<Library>>,
```

Comment:
does this really need a `Vec` of `Rc<Library>`? `ExtensionProxy` contains a `Rc<Library>` too.